src/pyscipopt/recipes/primal_dual_evolution.py

```diff
@@ -0,0 +1,60 @@
+    """
+    Attaches an event handler to a given SCIP model that collects primal and dual solutions,
+    along with the solving time when they were found.
+    The data is saved in model.data["primal_log"] and model.data["dual_log"]. They consist of
+    a list of tuples, each tuple containing the solving time and the corresponding solution.
+
+    A usage example can be found in examples/finished/plot_primal_dual_evolution.py. The
+    example takes the information provided by this recipe and uses it to plot the evolution
+    of the dual and primal bounds over time. 
+    """
+    class GapEventhdlr(Eventhdlr):
+
+        def eventinit(self): # we want to collect best primal solutions and best dual solutions
+            self.model.catchEvent(SCIP_EVENTTYPE.BESTSOLFOUND, self)
+            self.model.catchEvent(SCIP_EVENTTYPE.LPSOLVED, self)
+            self.model.catchEvent(SCIP_EVENTTYPE.NODESOLVED, self)
+            
+
+        def eventexec(self, event):
+            # if a new best primal solution was found, we save when it was found and also its objective
+            if event.getType() == SCIP_EVENTTYPE.BESTSOLFOUND:
+                self.model.data["primal_log"].append((self.model.getSolvingTime(), self.model.getPrimalbound()))
+            
+            if not self.model.data["dual_log"]:
+                self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+            
+            if self.model.getObjectiveSense() == "minimize":
+                if self.model.isGT(self.model.getDualbound(), self.model.data["dual_log"][-1][1]):
+                    self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+            else:
+                if self.model.isLT(self.model.getDualbound(), self.model.data["dual_log"][-1][1]):
+                    self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+
+        def eventexitsol(self):
+            if self.model.data["primal_log"][-1] and self.model.getPrimalbound() != self.model.data["primal_log"][-1][1]:
+                self.model.data["primal_log"].append((self.model.getSolvingTime(), self.model.getPrimalbound()))
+            
+            if not self.model.data["dual_log"]:
+                self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+            
+            if self.model.getObjectiveSense() == "minimize":
+                if self.model.isGT(self.model.getDualbound(), self.model.data["dual_log"][-1][1]):
+                    self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+            else:
+                if self.model.isLT(self.model.getDualbound(), self.model.data["dual_log"][-1][1]):
+                    self.model.data["dual_log"].append((self.model.getSolvingTime(), self.model.getDualbound()))
+
+
+    if not hasattr(model, "data"):
```

Comment:
That's a good point, but I'm not super sure what the best approach is. Personally, I like having a data dictionary in my model to store everything, so that it's easily accessible everywhere, but I understand that having things tidier might be preferable.

Reply:
This was just a question out of curiosity. I'm not against storing it in the `Model` object. I just think it's tidier and more intuitive to be in the event handler (this does not seem to be universal though). Happy with the current state.